### PR TITLE
[Kotlin Migration] IFieldController

### PR DESCRIPTION
**Pull request by Akshay0701**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt [hunk 1]
```diff
@@ -41,7 +41,7 @@ class BasicAudioClipFieldController : FieldControllerBase(), IFieldController {
 
     @KotlinCleanup("Convert to FixedTextView and make lateinit")
     private var tvAudioClip: TextView? = null
-    override fun createUI(context: Context, layout: LinearLayout) {
+    override fun createUI(context: Context?, layout: LinearLayout?) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:44]: `@KotlinCleanup("make context non-null")`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt [hunk 2]
```diff
@@ -64,7 +64,7 @@ class BasicAudioClipFieldController : FieldControllerBase(), IFieldController {
             val chooserPrompt = mActivity.resources.getString(R.string.multimedia_editor_popup_audio_clip)
             mActivity.startActivityForResultWithoutAnimation(Intent.createChooser(i, chooserPrompt), ACTIVITY_SELECT_AUDIO_CLIP)
         }
-        layout.addView(btnLibrary, ViewGroup.LayoutParams.MATCH_PARENT)
+        layout!!.addView(btnLibrary, ViewGroup.LayoutParams.MATCH_PARENT)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:67]: Is this required?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt [hunk 3]
```diff
@@ -75,10 +75,10 @@ class BasicAudioClipFieldController : FieldControllerBase(), IFieldController {
         layout.addView(tvAudioClip, ViewGroup.LayoutParams.MATCH_PARENT)
     }
 
-    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:78]: Liekwise, cleanup

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt [hunk 4]
```diff
@@ -35,49 +34,41 @@ import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity;
  * <p>
  * MultimediaEditFieldActivity calls controller's set methods by protocol before it works on UI creation.
  */
-public interface IFieldController {
+interface IFieldController {
     // This is guaranteed to be called before create UI, so that the controller
     // is aware of the field, including type an content.
-    void setField(IField field);
-
+    fun setField(field: IField?)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt:40]: Cleanup: non-null

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This seems fine, but needs a second thorough review

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt (hunk 2 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:67]: yes layout from parameter can be null, and its give error if i remove `!!`

**Review by david-allison-1 — APPROVED** [review]
LGTM, but thorough second review (also could you rebase relevant PRs after this is in)

**Comment by Akshay0701** [comment]
yes sure

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt (hunk 3 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:79]: This one appears to have a KotlinCleanup annotation to match the comment now, resolving

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.kt:44]: @Akshay0701 it looks we still need a KotlinCleanup annotation noting future desire to have context be non-null?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt:42]: has annotation now to match request, resolving

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt [hunk 5]
```diff
@@ -35,49 +35,43 @@ import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity;
...
+    @KotlinCleanup("make context non-null")
     // Layout is vertical inside a scroll view already
-    void createUI(Context context, LinearLayout layout);
-
+    fun createUI(context: Context?, layout: LinearLayout?)
 
     // If the controller ever starts an activity for result, this is going to be
     // called back on result.
-    void onActivityResult(int requestCode, int resultCode, Intent data);
+    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?)
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.kt:67]: I was curious if the intent could be non-null here so I checked. setResult (which feeds this) has two styles and one of them does not supply an Intent. So this has to be nullable.

TL;DR: this looks fine to me

**Review by mikehardy — CHANGES_REQUESTED** [review]
This looks really close to good - I gave it a good look as requested by David. I think there's one requested annotation still missing, and no other locations I could see that appeared to need them

**Review by mikehardy — APPROVED** [review]
